app/react-native/src/preview/components/OnDeviceUI/OnDeviceUI.tsx

```diff
@@ -136,49 +137,54 @@ const OnDeviceUI = ({
+    <>
+      <WrapperView style={[flex, IS_ANDROID && IS_EXPO && styles.expoAndroidContainer]}>
+        <KeyboardAvoidingView
+          enabled={!shouldDisableKeyboardAvoidingView || tabOpen !== PREVIEW}
+          behavior={IS_IOS ? 'padding' : null}
+          keyboardVerticalOffset={keyboardAvoidingViewVerticalOffset}
+          style={flex}
         >
-          <Animated.View style={previewWrapperStyles}>
-            <Animated.View style={previewStyles}>
-              <Preview disabled={tabOpen === PREVIEW}>
-                <StoryView story={story} />
-              </Preview>
-              {tabOpen !== PREVIEW ? (
-                <TouchableOpacity
-                  style={absolutePosition}
-                  onPress={() => handleToggleTab(PREVIEW)}
-                />
-              ) : null}
-            </Animated.View>
-          </Animated.View>
-          <Panel
-            style={getNavigatorPanelPosition(animatedValue.current, previewDimensions.width, wide)}
+          <AbsolutePositionedKeyboardAwareView
+            onLayout={setPreviewDimensions}
+            previewDimensions={previewDimensions}
           >
-            <StoryListView storyStore={storyStore} selectedStory={story} />
-          </Panel>
-          <Panel
-            style={getAddonPanelPosition(animatedValue.current, previewDimensions.width, wide)}
-          >
-            <Addons active={tabOpen === ADDONS} />
-          </Panel>
-        </AbsolutePositionedKeyboardAwareView>
-        <Navigation
-          tabOpen={tabOpen}
-          onChangeTab={handleToggleTab}
-          initialUiVisible={!isUIHidden}
-        />
-      </KeyboardAvoidingView>
-    </SafeAreaView>
+            <Animated.View style={previewWrapperStyles}>
+              <Animated.View style={previewStyles}>
+                <Preview disabled={tabOpen === PREVIEW}>
+                  <StoryView story={story} />
+                </Preview>
+                {tabOpen !== PREVIEW ? (
+                  <TouchableOpacity
+                    style={absolutePosition}
+                    onPress={() => handleToggleTab(PREVIEW)}
+                  />
+                ) : null}
+              </Animated.View>
+            </Animated.View>
+            <Panel
+              style={getNavigatorPanelPosition(
+                animatedValue.current,
+                previewDimensions.width,
+                wide
+              )}
+            >
+              <StoryListView storyStore={storyStore} selectedStory={story} />
+            </Panel>
+            <Panel
+              style={getAddonPanelPosition(animatedValue.current, previewDimensions.width, wide)}
+            >
+              <Addons active={tabOpen === ADDONS} />
+            </Panel>
+          </AbsolutePositionedKeyboardAwareView>
+        </KeyboardAvoidingView>
+      </WrapperView>
+      <Navigation tabOpen={tabOpen} onChangeTab={handleToggleTab} initialUiVisible={!isUIHidden} />
```

Comment:
here is the other change. the navigation bar is now positioned absolutely, no need to mess with safe areas of the deviceui.